### PR TITLE
#409: add tests for customer sync

### PR DESCRIPTION
**Pull request by vladarefiev**
PR for https://github.com/timelesslounge/timelessis/issues/409:
- small changes in customer_sync
- added test for customer_sync
- small fix in test for location_sync

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#513](https://github.com/timelesslounge/timelessis/issues/513) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#513](https://github.com/timelesslounge/timelessis/issues/513) is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), here is [why](https://www.0crat.com/footprint/CF9LG1YCA/40ee92b4-bd6c-4438-aa8d-a2e1d4ea1374); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @emilianodellacasa[/z](https://www.0crat.com/u/emilianodellacasa) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by vladarefiev** [comment]
@amihaiemil ping ;)

**Comment by vladarefiev** [comment]
@amihaiemil 🤔

**Comment by amihaiemil** [comment]
@vladarefiev right! I saw your PR today in the morning and I forgot. Sorry! On it now...

**Comment by vladarefiev** [comment]
@amihaiemil Wow, wow, @amihaiemil  inda house 8-)

File: timeless/poster/tasks.py [hunk 1]
```diff
@@ -46,11 +46,11 @@ def sync_tables():
 def sync_customers():
     """
     Periodic task for fetching and saving tables from Poster
+    Docs - https://dev.joinposter.com/docs/api#clients-getclients
     """
     for poster_customer in __poster_api().customers():
         customer = DB.session(Customer).query.filter_by(
-            first_name=poster_customer["first_name"],
-            last_name=poster_customer["last_name"]
+            poster_id=poster_customer["client_id"],
```
Comment (amihaiemil) [inline timeless/poster/tasks.py]: @vladarefiev I don't think this change is ok here, right now. I am also reviewing [this](https://github.com/timelesslounge/timelessis/pull/522) PR (solution for #410) and the logic of the merge method is based on the fact that customers are identified by first and last name here.

File: tests/integration/sync/it_sync_location_test.py [hunk 2]
```diff
@@ -36,21 +36,23 @@ def test_sync_location(locations_mock, auth_mock, db_session):
...
-        "city": "Edge City",
-        "address": "Blvd. Kukulcan Km 9.5 #30, Plaza Forum",
-        "longitude": 21.1326063,
-        "latitude": -86.7473191,
-        "type": "L",
-        "status": "open",
-        "comment": "Nightclub from a famous movie"
-    }]
+    locations_mock.return_value = {
+        "response": [{
```
Comment (amihaiemil) [inline tests/integration/sync/it_sync_location_test.py]: @vladarefiev why is the ``response`` parent-key needed now?

File: tests/integration/sync/it_sync_customers_test.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        "delete": "0",
+        "ewallet": "0"
+    }
+
+    customers_mock.return_value = {
+        "response": [
+            poster_customer,
+        ]
+    }
+
```
Comment (amihaiemil) [inline tests/integration/sync/it_sync_customers_test.py]: @vladarefiev Shouldn't there be 2 mocks here? 1 for the original customer in our system and the one returned by poster. Then, call the sync method and assert that the 2 are in sync. 

At the moment, it seems to me this test only checks the communication with Poster, not the actual merging of an existing Customer with a Poster Customer. WDYT?

File: tests/integration/sync/it_sync_customers_test.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    sync_customers()
+
+    customer = Customer.query.get(poster_id=poster_customer["client_id"])
```
Comment (amihaiemil) [inline tests/integration/sync/it_sync_customers_test.py]: @vladarefiev see my comment bellow, I think we should keep the firstname and lastname as coordinates.

File: tests/integration/sync/it_sync_location_test.py (hunk 2 above)
Comment (vladarefiev) [reply tests/integration/sync/it_sync_location_test.py]: @amihaiemil I have noticed in docs - https://dev.joinposter.com/docs/api#clients-getclients, So, I have to fix it! ;)

File: tests/integration/sync/it_sync_customers_test.py (hunk 4 above)
Comment (vladarefiev) [reply tests/integration/sync/it_sync_customers_test.py]: @amihaiemil I had thought like you before. But now I have recognized that poster is main source and `poster_id` is main key

**Comment by vladarefiev** [comment]
@amihaiemil I have just answered some of questions, I will take rest of them to think about them

File: timeless/poster/tasks.py (hunk 1 above)
Comment (vladarefiev) [reply timeless/poster/tasks.py]: @amihaiemil Thanks, done

File: tests/integration/sync/it_sync_customers_test.py (hunk 3 above)
Comment (vladarefiev) [reply tests/integration/sync/it_sync_customers_test.py]: @amihaiemil Agree

**Comment by vladarefiev** [comment]
@amihaiemil thanks for your attention. fixed, check it out now pls.

File: timeless/poster/tasks.py [hunk 5]
```diff
@@ -48,9 +48,10 @@ def sync_customers():
     Periodic task for fetching and saving tables from Poster
     Docs - https://dev.joinposter.com/docs/api#clients-getclients
     """
-    for poster_customer in __poster_api().customers():
-        customer = DB.session(Customer).query.filter_by(
-            poster_id=poster_customer["client_id"],
+    for poster_customer in __poster_api().customers().get('response', []):
+        customer = Customer.query.filter_by(
+            first_name=poster_customer["firstname"],
+            last_name=poster_customer["lastname"]
```
Comment (amihaiemil) [inline timeless/poster/tasks.py]: @vladarefiev See comment from @emilianodellacasa [here](https://github.com/timelesslounge/timelessis/pull/522#discussion_r264058340). So, after all, the syncing should be done by poster_id <-> client_id, not first_name, last_name. 

I thought it was done in the PR where the discussion took place, but no, it's not done yet. Can you revert it back to your initial change? :D

File: timeless/customers/models.py [hunk 6]
```diff
@@ -44,7 +44,7 @@ def create_by_poster(cls, poster_customer: dict):
         Method should return Customer object with given data from
         poster_customer dict
         """
-        return Customer(
+        return cls(
```
Comment (amihaiemil) [inline timeless/customers/models.py]: @vladarefiev I would leave it as ``Customer`` here, it's clearer than ``cls``, I believe.

File: tests/integration/sync/it_sync_customers_test.py [hunk 7]
```diff
@@ -49,7 +48,68 @@ def test_sync_customers(customers_mock, auth_mock):
...
+@pytest.mark.skip(
+    "Skipped until puzzle below is not done")
+@unittest.mock.patch.object(Authenticated, 'auth')
+@unittest.mock.patch.object(Poster, 'customers')
+def test_sync_customers_with_merging_data(customers_mock, auth_mock):
+    """
+    @todo #409:30min Remove skipping decorator from this test.
+     Now def merge_data(model, poster_data, timelessis_data) func is not
+     working properly, pls fix it. After fixing this test should work
+    """
```
Comment (amihaiemil) [inline tests/integration/sync/it_sync_customers_test.py]: @vladarefiev Why is that function not working?

File: timeless/poster/tasks.py (hunk 5 above)
Comment (vladarefiev) [reply timeless/poster/tasks.py]: @amihaiemil done

File: tests/integration/sync/it_sync_customers_test.py (hunk 7 above)
Comment (vladarefiev) [reply tests/integration/sync/it_sync_customers_test.py]: @amihaiemil I didn't investigate it. 
`sqlalchemy.orm.exc.FlushError: New instance <Customer at 0x7fd5b606e4a8> with identity key`

File: timeless/customers/models.py (hunk 6 above)
Comment (vladarefiev) [reply timeless/customers/models.py]: @amihaiemil I didn't agree, but if you want I will do it

Comment (amihaiemil) [reply timeless/customers/models.py]: @vladarefiev Is there any difference in functionality? I guess not. If it's just for reading purposes, then let it be ``Customer`` pls :D

**Comment by vladarefiev** [comment]
@amihaiemil done

File: timeless/customers/models.py (hunk 6 above)
Comment (vladarefiev) [reply timeless/customers/models.py]: @amihaiemil done

**Comment by amihaiemil** [comment]
@rultor good to merge

**Comment by emilianodellacasa** [comment]
@rultor merge

**Comment by 0crat** [comment]
Job [#513](https://github.com/timelesslounge/timelessis/issues/513) is not in the agenda of @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), can't inspect

**Comment by 0crat** [comment]
Job was finished in 44 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
The job [#513](https://github.com/timelesslounge/timelessis/issues/513) is now out of scope

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @emilianodellacasa[/z](https://www.0crat.com/u/emilianodellacasa)

**Comment by ypshenychka** [comment]
@0crat quality good

**Comment by 0crat** [comment]
Order was finished, quality is "good": +25 point(s) just awarded to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil)

**Comment by 0crat** [comment]
Quality review completed: +6 point(s) just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)

**Pull request merged**